src/merkle/smt/mod.rs

```diff
@@ -287,20 +288,41 @@ pub(crate) trait SparseMerkleTree<const DEPTH: u8> {
             });
         }
 
+        let mut reverse_mutations = BTreeMap::new();
         for (index, mutation) in node_mutations {
             match mutation {
-                Removal => self.remove_inner_node(index),
-                Addition(node) => self.insert_inner_node(index, node),
+                Removal => {
+                    if let Some(node) = self.remove_inner_node(index) {
+                        reverse_mutations.insert(index, Addition(node));
+                    }
+                },
+                Addition(node) => {
+                    if let Some(old_node) = self.insert_inner_node(index, node) {
+                        reverse_mutations.insert(index, Addition(old_node));
```

Comment:
Here we rely on the computation of mutations set, that it didn't generate useless update where new value is the same as previous (actually it does generate unnecessary changes in mutations set). Otherwise we would need to clone `node` and compare it with `old_node`.

Reply:
> actually it does generate unnecessary changes in mutations set

Does this mean we could improve `compute_mutations()` method?

Reply:
Yes, you're right, we could improve it in order to filter out unnecessary mutations (change value from `A` to `A`).